apps/app_voicemail.c

```diff
@@ -11413,6 +11412,365 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
+				context = vmu->context;
+			}
+		}
+		AST_LIST_UNLOCK(&users);
+	} else if (pos == 4 || pos == 8 || (pos == 6 && maxpos == 6) ) {
+		/* Walk through the standard folders */
+		wordlen = strlen(word);
+		for (int i = 0; i < ARRAY_LEN(mailbox_folders); i++) {
+			if (folder && !strncasecmp(word, mailbox_folders[i], wordlen) && ++which > state) {
+				return ast_strdup(mailbox_folders[i]);
+			}
+			folder = mailbox_folders[i];
+		}
+	} else if (pos == 5) {
+		/* find messages in the folder */
+		struct ast_vm_mailbox_snapshot *mailbox_snapshot;
+		struct ast_vm_msg_snapshot *msg;
+		int i = 0;
+		mailbox = a->argv[2];
+		context = a->argv[3];
+		folder = a->argv[4];
+		wordlen = strlen(word);
+
+		/* Take a snapshot of the mailbox and snag the individual info */
+		if ((mailbox_snapshot = ast_vm_mailbox_snapshot_create(mailbox, context, folder, 0, AST_VM_SNAPSHOT_SORT_BY_ID, 0))) {
+			/* we are only requesting the one folder, but we still need to know it's index */
+			for (i = 0; i < ARRAY_LEN(mailbox_folders); i++) {
+				if (!strcasecmp(mailbox_folders[i], folder)) {
+					break;
+				}
+			}
+			AST_LIST_TRAVERSE(&((mailbox_snapshot)->snapshots[i]), msg, msg) {
+				if (id && !strncasecmp(word, msg->msg_id, wordlen) && ++which > state) {
+					ret = ast_strdup(msg->msg_id);
+					break;
+				}
+				id = msg->msg_id;
+			}
+			/* done, destroy. */
+			mailbox_snapshot = ast_vm_mailbox_snapshot_destroy(mailbox_snapshot);
+		} else {
+			ast_log(LOG_NOTICE, "Can't create snapshot for voicemail box %s@%s\n", mailbox, context);
+		}
+	}
+
+	return ret;
+}
+
+static char *handle_voicemail_forward_message(struct ast_cli_entry *e, int cmd, struct ast_cli_args *a)
+{
+	switch (cmd) {
+	case CLI_INIT:
+		e->command = "voicemail forward";
+		e->usage =
+		"Usage: voicemail forward <from_mailbox> <from_context> <from_folder> <messageid> <to_mailbox> <to_context> <to_folder>\n"
+		"       Forward message <messageid> in mailbox <mailbox>@<context> <from_folder>\n"
+		"       to mailbox <mailbox>@<context> <to_folder>\n";
+		return NULL;
+	case CLI_GENERATE:
+		return complete_voicemail_move_message(a, 8);
+	case CLI_HANDLER:
+		break;
+	}
+
+	if (a->argc != 9) {
+		return CLI_SHOWUSAGE;
+	}
+
+	if (forward_message_from_mailbox(a)){
+		ast_cli(a->fd, "Error forwarding message!\n");
+	}
+
+	return CLI_SUCCESS;
+}
+
+static char *handle_voicemail_move_message(struct ast_cli_entry *e, int cmd, struct ast_cli_args *a)
+{
+	switch (cmd) {
+	case CLI_INIT:
+		e->command = "voicemail move";
+		e->usage =
+		"Usage: voicemail move <mailbox> <context> <from_folder> <messageid> <to_folder>\n"
+		"       Move message <messageid> in mailbox <mailbox>&<context> from <from_folder> to <to_folder>\n";
+		return NULL;
+	case CLI_GENERATE:
+		return complete_voicemail_move_message(a, 6);
+	case CLI_HANDLER:
+		break;
+	}
+
+	if (a->argc != 7) {
+		return CLI_SHOWUSAGE;
+	}
+
+	if (move_message_from_mailbox(a)){
+		ast_log(AST_LOG_WARNING, "Error moving message!\n");
```

Comment:
Probably should be an error, not a warning, and it should include details of what message move failed.

Reply:
Modified all three commands to output pass/fail with details.